### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by bollwyvl**
This adds the test harness from nbpresent for end-to-end integration tests.

It uses nose and casperjs, building on the notebook's `jstest.py`. I am behind the ball on the 4.0.6 -> 4.1 transition, but i did leave myself some cryptic comments about it, so I should be alright.

My initial plan is to get the tests working locally with local `anaconda login` privileges. Will continue!

### CODE REVIEW

**Comment by bollwyvl** [comment]
Just updated with better screenshot naming to help track down timing issues (one directory per test).

The more issues I encounter, the more I think I have to `mock` anaconda.org's requests as fixtures... should be relatively easy, make it testable on anaconda.org, and give us better control over our tests. in return, we'll have to make sure that the output is easily generateable, and get on a refresh schedule to regenerate them... i guess tracking to the `anaconda-client` version.

**Comment by bollwyvl** [comment]
Getting somewhere now! Turns out the routes weren't resolving for two reasons:
- the extension was not installing correctly in the "pristine" (har) test environment
  - for the test I am overloading the `server_command`, but it could be that is unnecessary
- the tornado routes were not properly prefixed on the client
  - `"/ac-login/"` vs `Jupyter.notebook.base_url + "/ac-login/"`
  - no doubt other extensions with a JS component that talks to a custom server route will make this same mistake
  - this is important to ensure that extensions can be part of more nuanced deployments (e.g. binder)

at this point, it works against local creds. It may _still_ be of interest to do the appropriate mocking, if only so that the tests work no matter where they are run... once the environment is installed correctly, which is an additional piece of metadata that's missing... once all tests are run inside `conda build`, this will fall out naturally, and there's no point to doing a `requirements-dev.txt`, probably.

**Comment by bollwyvl** [comment]
Hooray, first working linux build, with passing tests, on anaconda cloud!

http://anaconda.org/nbcio/nb_anacondacloud/builds/7/0

:cake:

**Comment by bollwyvl** [comment]
Ok, this is actually "testing" upload (using a fixture response for both package and release).

The coverage is still reporting pretty low, but it could be that it would be higher if the two test runs were combined, since one handles auth and the other no auth.

I still want to get back my local auth option, as I think it is obviously still valuable to do real integration testing, but having a stable set of fixtures were a key step.

**Comment by bollwyvl** [comment]
Added the local auth stuff back in, which will be good as a baseline to be used locally, complimented by the mocked integration testing.

By my estimation, unless we want to chase down coverage, this is ready to merge... I would definitely appreciate some feedback from @malev or suitable delegates to see if my approach is crazy.

**Comment by damianavila** [comment]
Uppsss this was automatically merged when I merge #6... weird...

**Comment by damianavila** [comment]
please @malev and others... review this one even if it is merged so @bollwyvl can have some feedback on this...

**Comment by damianavila** [comment]
> Uppsss this was automatically merged when I merge #6... weird...

I guess this is somewhat related with the rebase of the PR... but not sure... really weird.

**Comment by damianavila** [comment]
Yes, the other PR was rebased on this one, this is why it got merged automatically...

File: .binstar.yml [hunk 1]
```diff
@@ -2,21 +2,17 @@ package: nb_anacondacloud
...
   - python=3.5
 
-install:
-  - conda build conda.recipe -c anaconda-notebook
-  - conda install nb_anacondacloud --use-local -c anaconda-notebook
-
-test:
-  - python -m unittest discover
+script:
+  - conda build conda.recipe -c anaconda-notebook -c javascript -c mutirri -c jakirkham -c bokeh
```
Comment (damianavila) [inline .binstar.yml:16]: OK, post merge review :wink:

any package needed should be migrated/copy to the new org containing the extensions...

File: .binstar.yml [hunk 2]
```diff
@@ -2,21 +2,17 @@ package: nb_anacondacloud
...
 platform:
   - linux-64
-  - linux-32
-  - osx-64
-  - win-64
-  - win-32
+  # - linux-32
+  # - osx-64
+  # - win-64
+  # - win-32
```
Comment (damianavila) [inline .binstar.yml:8]: we need build in all the platforms... if running the test is difficult in win, we could potentially deactivate the testing on win until the problems are solved, but still, we should build in all the platforms, I think...

File: nb_anacondacloud/static/main.js [hunk 3]
```diff
@@ -25,7 +26,7 @@ define(['jquery', 'base/js/dialog'], function ($, dialog) {
         }
 
         $.ajax({
-            url: "/ac-publish",
+            url: Jupyter.notebook.base_url + "ac-publish",
```
Comment (damianavila) [inline nb_anacondacloud/static/main.js:29]: this is important and, in fact, impacted in other extensions: https://github.com/Anaconda-Server/nb_revisioncontrol/pull/5/files#diff-44ce3f32fb6ea0ec59591a50d801ff67R43

**Comment by damianavila** [comment]
BTW, sorry by the early merge... I should have been figured out you rebased...

**Pull request merged**